aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -92,18 +97,111 @@ impl SessionId {
+            let (modules, resources) = account_changeset.into_inner();
+
+            // * Separate the resource groups from the non-resource groups
+            // * non-resource groups are kept as is
+            // * resource groups are merged into the correct format as deltas to the source data
+            for (struct_tag, blob_op) in resources {
+                let resource_group = self.remote.get_resource_group(&struct_tag).map_err(|_| common_error.clone())?;
+                if let Some(resource_group) = resource_group {
+                    resource_groups
+                        .entry(resource_group)
+                        .or_insert_with(AccountChangeSet::new)
+                        .add_resource_op(struct_tag, blob_op)
+                        .map_err(|_| common_error.clone())?;
+                    continue;
+                }
+
+                account_changeset_grouped
+                    .add_resource_op(struct_tag, blob_op)
+                    .map_err(|_| common_error.clone())?;
+            }
+
+            for (resource_tag, resources) in resource_groups {
+                let source_data = self
+                    .remote
+                    .get_resource_group_data(&addr, &resource_tag)
+                    .map_err(|_| common_error.clone())?;
+                let (mut source_data, create) = if let Some(source_data) = source_data {
+                    let source_data =
+                        bcs::from_bytes(&source_data).map_err(|_| common_error.clone())?;
+                    (source_data, false)
+                } else {
+                    (BTreeMap::new(), true)
+                };
+
+                for (struct_tag, current_op) in resources.into_resources() {
+                    match current_op {
+                        MoveStorageOp::Delete => {
+                            source_data
+                                .remove(&struct_tag)
+                                .ok_or_else(|| common_error.clone())?;
+                        },
+                        MoveStorageOp::Modify(new_data) => {
+                            let data = source_data
+                                .get_mut(&struct_tag)
+                                .ok_or_else(|| common_error.clone())?;
+                            *data = new_data;
+                        },
+                        MoveStorageOp::New(data) => {
+                            let data = source_data.insert(struct_tag, data);
+                            if data.is_some() {
+                                return Err(common_error);
+                            }
+                        },
+                    }
+                }
+
+                let op = if source_data.is_empty() {
+                    MoveStorageOp::Delete
+                } else if create {
+                    MoveStorageOp::New(
+                        bcs::to_bytes(&source_data).map_err(|_| common_error.clone())?,
```

Comment:
Same question here about using common error (UNREACHABLE) everywhere

Reply:
if we can not serialize, something is very wrong...

Reply:
again an invariant.. I don't like expect or unwrap